modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/LocalSubscription.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.andes.subscription;
+
+import org.apache.commons.lang.builder.HashCodeBuilder;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.andes.configuration.AndesConfigurationManager;
+import org.wso2.andes.configuration.enums.AndesConfiguration;
+import org.wso2.andes.kernel.*;
+import org.wso2.andes.mqtt.MQTTLocalSubscription;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * This is the class that represents a subscription in andes kernel. It is responsible
+ * for handling both inbound (protocol > kernel) and outbound (kernel > protocol) subscription
+ * events. For handling outbound events it keeps a OutboundSubscription object and forward
+ * requests
+ */
+public class LocalSubscription  extends BasicSubscription implements InboundSubscription{
+
+    private static Log log = LogFactory.getLog(LocalSubscription.class);
+
+    /**
+     * Outbound subscription reference. We forward outbound events to this object. Get its response
+     * and act upon (make kernel side changes)
+     */
+    private OutboundSubscription subscription;
+
+    /**
+     * Map to track messages being sent <message id, MsgData reference>
+     */
+    private final ConcurrentHashMap<Long, DeliverableAndesMetadata> messageSendingTracker
+            = new ConcurrentHashMap<>();
+
+    /**
+     * Count sent but not acknowledged message count for channel of the subscriber
+     */
+    private AtomicInteger unAcknowledgedMsgCount = new AtomicInteger(0);
+
+    private Integer maxNumberOfUnAcknowledgedMessages = 100000;
```

Comment:
Can this value be changed after object creation. If not shall we make this final and initialize only through the AndesConfigurationManager?